graphistry/embed_utils.py

```diff
@@ -0,0 +1,593 @@
+import numpy as np
+import pandas as pd
+import torch
+import torch.nn as nn
+import dgl
+from dgl.dataloading import GraphDataLoader
+import torch.nn.functional as F
+from .networks import RGCNEmbed
+from tqdm import trange
+import logging
+logger = logging.getLogger(__name__)
+
+
+class EmbedDistScore:
+
+    @staticmethod
+    def TransE(h, r, t):
+        return (h + r - t).norm(p=1, dim=1)
+
+    @staticmethod
+    def DistMult(h, r, t):
+        return (h * r * t).sum(dim=1)
+
+    @staticmethod
+    def RotatE(h, r, t):
+        return -(h * r - t).norm(p=1, dim=1)
+
+
+class HeterographEmbedModuleMixin(nn.Module):
+    def __init__(self):
+        super().__init__()
+
+        self.protocol = {
+            'TransE': EmbedDistScore.TransE,
+            'DistMult': EmbedDistScore.DistMult,
+            'RotatE': EmbedDistScore.RotatE
+        }
+
+    def _preprocess_embedding_data(self, train_split=0.8):
+        print('Preprocessing embedding data')
+        src, dst = self._source, self._destination
+        relation = self._relation
+
+        if self._node is not None and self._nodes is not None:
+            nodes = self._nodes[self._node]
+        elif self._node is None and self._nodes is not None:
+            nodes = list(range(
```

Comment:
confusing to mix `List` with `Series`, let's stick with `Series`... (`self._nodes.reset_index(drop=True).reset_index().index` ?)

Reply:
should I replace all the lists to pandas series ?